src/app/archive/lib/processor.ml

```diff
@@ -0,0 +1,356 @@
+open Core
+open Async
+open Pipe_lib
+open Coda_base
+open Signature_lib
+
+let receiver user_command =
+  match (User_command.payload user_command).body with
+  | Payment payment ->
+      payment.receiver
+  | Stake_delegation (Set_delegate delegation) ->
+      delegation.new_delegate
+
+module type Data_intf = sig
+  type t [@@deriving compare]
+
+  include Comparable.S with type t := t
+end
+
+module type Graphql_obj_intf = sig
+  type t
+
+  type hash
+
+  val id : t -> int
+
+  val hash : t -> hash
+end
+
+module type Inputs_intf = sig
+  module Hash : Data_intf
+
+  module Data : sig
+    type t
+  end
+
+  module Graphql_query_result : Graphql_obj_intf with type hash := Hash.t
+
+  val get_existing :
+    Hash.t array -> Graphql_query_result.t array Deferred.Or_error.t
+
+  val update :
+       Data.t Hash.Map.t
+    -> Graphql_query_result.t Hash.Map.t
+    -> unit Deferred.Or_error.t
+
+  val insert :
+       Data.t Hash.Map.t
+    -> Graphql_query_result.t array option Deferred.Or_error.t
+end
+
+module Make_upsert (Inputs : Inputs_intf) = struct
+  open Inputs
+
+  let upsert (hash_to_data_map : Data.t Hash.Map.t) =
```

Comment:
We can't really use Hasura's upsert functionalities. This is mainly because one of the data fields for graphql arguments to do an upsert have the method `constraint`. This is keyword in ocaml and we can't really create an object with the field, constraint. So, this prevents us from doing an upsert.

One way to overcome this issue is by creating a ppx that allows you to create an object with a field of any name. I will probably do this later to make it possible to do some type of upsert.

Reply:
I don't understand -- you used `constraint` in the insert graphql above?

```
    on_conflict: {constraint: user_commands_pkey, update_columns: first_seen}
```

Does `constraint` have to be used differently here?

It's a bit scary to land this just in case we forget to fix it. But if it is a fix we can do by updating the graphql ppx somehow maybe if we stuck a TODO for @mrmr1993 (as Matt already is comfortable with the graphql ppx internals)

Reply:
@bkase Yes `constraint` has to be used different for public keys in User_commands. Namely, for the user_command input argument, there is a nested type for doing upserts for the nested public keys. Namely, the type would look like something like the following:

```
 on_conflict: {constraint: public_key_pkey, update_columns: []}
```

However, it is impossible to create a user_command with this nested objected because `constraint` is keyword in ocaml.

Reply:
I do have an awesome hack that addresses this issue. Essentially, after getting the introspective query from Hasura for GraphQL, we can create a python script that replaces all the usages of `constraint` with `_constraint`. Then, we can easily create this `on_conflict` object with the field `_constraint`. 

Afterwards, we can feed this object into graphql_ppx. It will generate the json inputs for the graphql query. We can then convert all the `_constraint` fields back to `constraint` and perform the actual graphql query!

Reply:
If we had some time, we would probably augment graphql_ppx to detect keywords and replace them with `keyword_`, aka do this same swap logic but inside graphql_ppx to make it a little less error-prone/manual. This seems like it could be a ok workaround for now though

Reply:
Okay this issue is resolved. I made the script to resolve this issue in OCaml.